server/sonar-server/src/main/java/org/sonar/server/qualityprofile/DefinedQProfileRepositoryImpl.java

```diff
@@ -0,0 +1,200 @@
+   *       with a given name has {@link RulesProfile#getDefaultProfile()} is {@code true}</li>
+   *   <li>{@link DefinedQProfile.Builder#activeRules activeRules}: the concatenate of the active rules of all
+   *       RulesProfile with a given name</li>
+   * </ul>
+   */
+  private static List<DefinedQProfile.Builder> toQualityProfileBuilders(Map.Entry<String, List<RulesProfile>> rulesProfilesByLanguage) {
+    String language = rulesProfilesByLanguage.getKey();
+    // use a LinkedHashMap to keep order of insertion of RulesProfiles
+    Map<String, DefinedQProfile.Builder> qualityProfileBuildersByName = new LinkedHashMap<>();
+    for (RulesProfile rulesProfile : rulesProfilesByLanguage.getValue()) {
+      qualityProfileBuildersByName.compute(
+        rulesProfile.getName(),
+        (name, existingBuilder) -> updateOrCreateBuilder(language, existingBuilder, rulesProfile, name));
+    }
+    return ImmutableList.copyOf(qualityProfileBuildersByName.values());
+  }
+
+  /**
+   * Fails if more than one {@link DefinedQProfile.Builder#declaredDefault} is {@code true}, otherwise returns {@code true}.
+   */
+  private static boolean ensureAtMostOneDeclaredDefault(Map.Entry<String, List<DefinedQProfile.Builder>> entry) {
+    Set<String> declaredDefaultProfileNames = entry.getValue().stream()
+      .filter(DefinedQProfile.Builder::isDeclaredDefault)
+      .map(DefinedQProfile.Builder::getName)
+      .collect(Collectors.toSet());
+    checkState(declaredDefaultProfileNames.size() <= 1, "Several Quality profiles are flagged as default for the language %s: %s", entry.getKey(), declaredDefaultProfileNames);
+    return true;
+  }
+
+  private static DefinedQProfile.Builder updateOrCreateBuilder(String language, @Nullable DefinedQProfile.Builder existingBuilder, RulesProfile rulesProfile, String name) {
+    DefinedQProfile.Builder builder = existingBuilder;
+    if (builder == null) {
+      builder = new DefinedQProfile.Builder()
+        .setLanguage(language)
+        .setName(name);
+    }
+    Boolean defaultProfile = rulesProfile.getDefaultProfile();
+    boolean declaredDefault = defaultProfile != null && defaultProfile;
+    return builder
+      // if there is multiple RulesProfiles with the same name, if at least one is declared default,
+      // then QualityProfile is flagged as declared default
+      .setDeclaredDefault(builder.isDeclaredDefault() || declaredDefault)
+      .addRules(rulesProfile.getActiveRules());
+  }
+
+  private static List<DefinedQProfile> toQualityProfiles(List<DefinedQProfile.Builder> builders) {
+    if (builders.stream().noneMatch(DefinedQProfile.Builder::isDeclaredDefault)) {
+      Optional<DefinedQProfile.Builder> sonarWayProfile = builders.stream().filter(builder -> builder.getName().equals(DEFAULT_PROFILE_NAME)).findFirst();
+      if (sonarWayProfile.isPresent()) {
+        sonarWayProfile.get().setComputedDefault(true);
+      } else {
+        builders.iterator().next().setComputedDefault(true);
```

Comment:
That works because algorithm starts from ProfileDefinition but not Languages. It's not straightforward to understand. I suggest to verify that there's at least one element defined before calling `setComputeDefault()`.